keel-igor/src/main/kotlin/com/netflix/spinnaker/keel/igor/BuildLifecycleMonitor.kt

```diff
@@ -0,0 +1,197 @@
+import com.netflix.spinnaker.keel.api.artifacts.BuildMetadata
+import com.netflix.spinnaker.keel.front50.Front50Service
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.FAILED
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.RUNNING
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.SUCCEEDED
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.UNKNOWN
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventType
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventType.BUILD
+import com.netflix.spinnaker.keel.lifecycle.LifecycleMonitor
+import com.netflix.spinnaker.keel.lifecycle.LifecycleMonitorRepository
+import com.netflix.spinnaker.keel.lifecycle.MonitoredTask
+import com.netflix.spinnaker.keel.igor.artifact.ArtifactMetadataService
+import com.netflix.spinnaker.keel.lifecycle.LifecycleEventStatus.ABORTED
+import com.netflix.spinnaker.keel.retrofit.isNotFound
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.context.properties.EnableConfigurationProperties
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+import retrofit2.HttpException
+
+@Component
+@EnableConfigurationProperties(LifecycleConfig::class)
+class BuildLifecycleMonitor(
+  override val monitorRepository: LifecycleMonitorRepository,
+  override val publisher: ApplicationEventPublisher,
+  override val lifecycleConfig: LifecycleConfig,
+  val objectMapper: ObjectMapper,
+  val artifactMetadataService: ArtifactMetadataService,
+  val front50Service: Front50Service,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String
+) : LifecycleMonitor(monitorRepository, publisher, lifecycleConfig) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override fun handles(type: LifecycleEventType): Boolean =
+    type == BUILD
+
+  override suspend fun monitor(task: MonitoredTask) {
+    val buildData = parseAndValidate(task) ?: return
+    kotlin.runCatching {
+      artifactMetadataService.getArtifactMetadata(buildData.buildNumber, buildData.commitId)
+    }.onSuccess { metadata ->
+      val buildMetadata = metadata?.buildMetadata
+      if (buildMetadata == null) {
+        log.error("Error fetching status for $task, response was null")
+        handleFailureFetching(task)
+      } else {
+        when (buildMetadata.status) {
+          "BUILDING" -> publishRunningEvent(task)
+          "SUCCESS" -> publishSucceededEvent(task, buildMetadata)
+          "FAILURE" -> publishFailedEvent(task, buildMetadata)
+          "ABORTED" -> publishAbortedEvent(task, buildMetadata)
+          "UNSTABLE" -> publishSucceededEvent(task, buildMetadata)
+          else -> publishUnknownStatusEvent(task, buildMetadata.status)
+        }
+
+        if (buildMetadata.isComplete()) {
+          endMonitoringOf(task)
```

Comment:
Same comment on making method names self-contained.

Reply:
Or even make it an extension function so it reads as `task.endMonitoring()` 🤯